server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -1487,4 +1529,48 @@ private void setIdFieldDataEnabled(boolean value) {
         }
         return Optional.empty();
     }
+
+    private void updateDanglingIndicesInfo(Index index) {
+        assert DiscoveryNode.isDataNode(settings) : "dangling indices information should only be persisted on data nodes";
+        if (danglingIndicesToWrite.add(index)) {
+            logger.trace("triggered dangling indices update for {}", index);
+            final long triggeredTimeMillis = threadPool.relativeTimeInMillis();
+            try {
+                danglingIndicesThreadPoolExecutor.execute(new AbstractRunnable() {
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.info(() -> new ParameterizedMessage("failed to write dangling indices state for index {}", index), e);
+                    }
+
+                    @Override
+                    protected void doRun() {
+                        final boolean exists = danglingIndicesToWrite.remove(index);
```

Comment:
Doing this before calling `indexService.writeDanglingIndicesInfo()` means that `allPendingDanglingIndicesWritten()` may return `true` while there's still a write in progress.

Reply:
good catch. In an earlier iteration, I had `allPendingDanglingIndicesWritten` checking the executor as well, but I somehow thought this was unneeded later. I've reinstated the check in 3a7b656f5ce